src/concurrency/thread.rs

```diff
@@ -253,6 +288,22 @@ impl<'mir, 'tcx> Default for ThreadManager<'mir, 'tcx> {
     }
 }
 
+impl VisitMachineValues for ThreadManager<'_, '_> {
+    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>)) {
+        let ThreadManager { threads, thread_local_alloc_ids, .. } = self;
+
+        for thread in threads {
+            thread.visit_machine_values(visit);
+        }
+        for ptr in thread_local_alloc_ids.borrow().values().copied() {
+            let ptr: Pointer<Option<Provenance>> = ptr.into();
+            visit(&Operand::Indirect(MemPlace::from_ptr(ptr)));
+        }
+        // FIXME: Do we need to do something for TimeoutCallback? That's a Box<dyn>, not sure what
```

Comment:
Oh, good catch. That's tricky to handle -- if we let callbacks capture things (and I assume they do capture something) then we either have to declare that they must not capture any machine values (which is tricky to check) or we have to replace the closures by an explicit struct + trait.

Reply:
I don't know if what I implemented is exactly what you had in mind, but I ended up with a surprised lifetime bound. I suspect it comes from the fact that I'm not using `FnOnce`? I suppose it's possible I also got a `'mir` and `'tcx` backwards but I've been over the code a bunch of times now :sleeping:

Reply:
Oli got me straightened out here, still not entirely sure of the implementation.

Reply:
Yes that looks like the struct+trait solution. :)